### PR TITLE
Day20

### PR DESCRIPTION
**Pull request by mtsochacki**

### CODE REVIEW

File: 2021/day20/day20.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            sc.close();
+        } catch (Exception e) {
+            System.out.println("Something went wrong " + e);
+        }
+        return image;
+    }
+
+    public static ArrayList<ArrayList<Character>> surroundWithDots(ArrayList<ArrayList<Character>> image) {
+        ArrayList<ArrayList<Character>> newImage = new ArrayList<>();
+        newImage.add(new ArrayList<>());
```
Comment (Lammatian) [inline 2021/day20/day20.java:37]: This should be replaced with `new ArrayList<>(Collections.nCopies(image.get(0).size() + 2), '.')`

File: 2021/day20/day20.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            newLine.add('.');
+            newLine.addAll(line);
+            newLine.add('.');
+            newImage.add(newLine);
+        }
+        newImage.add(new ArrayList<>());
+        for (int i = 0; i < newImage.get(1).size(); i++) {
+            newImage.get(0).add('.');
+            newImage.get(newImage.size() - 1).add('.');
+        }
```
Comment (Lammatian) [inline 2021/day20/day20.java:49]: This should be replaced with `new ArrayList<>(Collections.nCopies(image.get(0).size() + 2), '.')` (along with line 37)

File: 2021/day20/day20.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    public static ArrayList<Character> stringToList(String line) {
+        ArrayList<Character> result = new ArrayList<>();
+        for (int i = 0; i < line.length(); i++) {
+            result.add(line.charAt(i));
+        }
+        return result;
+    }
+
+    public static int getIndex(ArrayList<ArrayList<Character>> image, int x, int y) {
```
Comment (Lammatian) [inline 2021/day20/day20.java:61]: Don't really like the naming. Maybe `determineOutputPixelIndex`?

File: 2021/day20/day20.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        for (int i = -1; i <= 1; i++) {
+            for (int j = -1; j <= 1; j++) {
+                stringIndex.append(image.get(y + i).get(x + j));
+            }
+        }
+        String binaryString = stringIndex.toString().replace("#", "1").replace(".", "0");
+        return Integer.parseInt(binaryString, 2);
+    }
+
+    public static int countPixels(int steps) {
```
Comment (Lammatian) [inline 2021/day20/day20.java:72]: `countPixels` -> `countLightPixels`

File: 2021/day20/day20.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                image.add(stringToList(sc.nextLine()));
+            }
+            sc.close();
+        } catch (Exception e) {
+            System.out.println("Something went wrong " + e);
+        }
+        return image;
+    }
+
+    public static ArrayList<ArrayList<Character>> surroundWithDots(ArrayList<ArrayList<Character>> image) {
```
Comment (Lammatian) [inline 2021/day20/day20.java:35]: This smells like a (theoretically) great place to use linked lists - you keep on adding things at the beginning and at the end. However, do note that linked lists may not be faster than dynamic arrays in practice, even if they are in theory - read up on cache locality if you're interested in this.

On the other hand, you do need the indices when you do the checking of each pixel, so it's difficult to tell which approach is faster. How about some small benchmarking again?

File: 2021/day20/day20.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            }
+        }
+        String binaryString = stringIndex.toString().replace("#", "1").replace(".", "0");
+        return Integer.parseInt(binaryString, 2);
+    }
+
+    public static int countPixels(int steps) {
+        String algorithm = readAlgorithm("data.txt");
+        ArrayList<ArrayList<Character>> image = readImage("data.txt");
+        int var = steps * 2;
```
Comment (Lammatian) [inline 2021/day20/day20.java:75]: I almost threw up (just use `2 * steps` in the for loop condition if you're already at this arcane solution)

File: 2021/day20/day20.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        for (int i = 0; i < var; i++) {
```
Comment (Lammatian) [inline 2021/day20/day20.java:76]: You can add `2 * steps + 1` 'border lines' at the beginning of the algorithm and then not worry about adding more later. Thus line 80 is spurious if you add one more line here. Also this should speed things up every so slightly.

File: 2021/day20/day20.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+            image = surroundWithDots(image);
+        }
+        for (int i = 0; i < steps; i++) {
+            image = surroundWithDots(image);
```
Comment (Lammatian) [inline 2021/day20/day20.java:80]: Unnecessary per earlier comment

File: 2021/day20/day20.java [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+            ArrayList<ArrayList<Character>> tmpImage = new ArrayList<>();
+            for (ArrayList<Character> line : image) {
+                ArrayList<Character> tmpLine = new ArrayList<>(line);
+                tmpImage.add(tmpLine);
+            }
```
Comment (Lammatian) [inline 2021/day20/day20.java:85]: You don't need to initialize `tmpImage` with `image`, you can just make an empty array of arrays of appropriate size.

File: 2021/day20/day20.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            for (int y = 1; y < tmpImage.size() - 1; y++) {
+                for (int x = 1; x < tmpImage.get(0).size() - 1; x++) {
+                    tmpImage.get(y).set(x, algorithm.charAt(getIndex(image, x, y)));
```
Comment (Lammatian) [inline 2021/day20/day20.java:88]: the `algorithm.charAt(getIndex(image, x, y))` could be a separate function `getOutputPixel`

**Review by Lammatian — CHANGES_REQUESTED** [review]
Quite a few corrections. Still can't get over `int var`. As discussed in DMs, consider adjusting the algorithm altogether - either by modifying the 'surrounding' to do it all at first or by surrounding after each step, but with correct type of pixel.

**Comment by mtsochacki** [comment]
Lot's of things to change, but I am merging for now

**Pull request merged**